crucible/src/Lang/Crucible/Simulator/VecValue.hs

```diff
@@ -0,0 +1,492 @@
+import Lang.Crucible.Types
+import Lang.Crucible.Backend
+import Lang.Crucible.Simulator.SimError
+
+-- | We expect this to have one instance: for `RegValue'`.
+-- We don't use that type directly to avoid recursive module dependencies.
+class IsExprBuilder sym => IsRegValue sym f | f -> sym where
+
+  -- | Convert a partial expression into a value of a @Maybe@ type.
+  partialToMaybe :: PartExpr (Pred sym) (f tp) -> f (MaybeType tp)
+
+-- | We expect this to have two instances:
+-- One for 'IO', used by the symbolic evaluator,
+-- and one for 'MuxLeafT' used by `crucible-mir`
+class MonadIO m => VecMonad sym m where
+
+  -- | We use this for partial operations (e.g., indexing a vector out of bounds).
+  vecAbort :: IsSymBackend sym bak => bak -> SimErrorReason -> m a
+
+  -- | A pre-condition on the current operation.  Assertions we generate
+  -- are conditional on this.
+  vecPre   :: IsExprBuilder sym => sym -> m (Pred sym)
+
+instance VecMonad sym IO where
+  vecAbort = addFailedAssertion 
+  vecPre = pure . truePred
+
+-- | Type suitable for indexing into vectors, or getting the size of a vector.
+-- Instances:
+--  * one for symbolic natural numbers,
+--  * one for bit-vector types (i.e., usize),
+-- * one for concrete natural numbers
+class IsExprBuilder sym => VecSize sym t where
+  vecSizeIsConcrete :: sym -> t -> Maybe Int
+  vecSizeLit        :: sym -> Natural -> IO t
+  vecSizeEq         :: sym -> t -> t -> IO (Pred sym)
+  vecSizeLeq        :: sym -> t -> t -> IO (Pred sym)
+
+
+
+instance IsExprBuilder sym => VecSize sym (SymNat sym) where
+  vecSizeIsConcrete _ i =
+    do
+      n <- asNat i
+      guard (n <= fromIntegral (maxBound :: Int))
+      pure (fromIntegral n)
+
+  vecSizeLit sym i = natLit sym (fromIntegral i)
+  vecSizeEq        = natEq
+  vecSizeLeq       = natLe
+
+instance (1 <= w, KnownNat w, IsExprBuilder sym, SymBV sym w ~ BVExpr s w) => VecSize sym (BVExpr s w) where
+  vecSizeIsConcrete _ i =
+    do BV.BV n <- asBV i
+       guard (n <= toInteger (maxBound :: Int))
+       pure (fromIntegral n)
+  vecSizeLit sym n = bvLit sym knownNat (BV.mkBV knownNat (toInteger n))
+  vecSizeEq        = bvEq
+  vecSizeLeq       = bvUle
+
+instance IsExprBuilder sym => VecSize sym Natural where
+  vecSizeIsConcrete _ n = Just (fromIntegral n) -- assumes it fits
+  vecSizeLit _ = pure
+  vecSizeEq sym x y = pure (backendPred sym (x == y))
+  vecSizeLeq sym x y = pure (backendPred sym (x <= y))
+
+
+instance IsExprBuilder sym => VecSize sym Integer where
+  vecSizeIsConcrete _ n = Just (fromIntegral n) -- assumes it fits
+  vecSizeLit _ = pure . fromIntegral -- assumes it fits
+  vecSizeEq sym x y = pure (backendPred sym (x == y))
+  vecSizeLeq sym x y = pure (backendPred sym (x <= y))
+
+
+
+--------------------------------------------------------------------------------
+
+
+
+data VecVal (f :: CrucibleType -> Type) (tp :: CrucibleType) where
+  VecVal :: V.Vector (f tp) -> VecVal f tp
```

Comment:
The name of this module is "`VecValue`", but this data type is named "`VecVal`". For consistency, it would be nice to use the same name throughout. My personal preference would be "`VecValue`", as it more closely aligns with the names of other types in Crucible (e.g., `RegValue` and `AtomValue`).